channelz/funcs.go

```diff
@@ -0,0 +1,583 @@
+func (s int64Slice) Less(i, j int) bool { return s[i] < s[j] }
+
+func copyMap(m map[int64]string) map[int64]string {
+	n := make(map[int64]string)
+	for k, v := range m {
+		n[k] = v
+	}
+	return n
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func (c *channelMap) GetTopChannels(id int64) ([]*ChannelMetric, bool) {
+	c.mu.RLock()
+	l := len(c.topLevelChannels)
+	ids := make([]int64, 0, l)
+	cns := make([]*channel, 0, min(l, EntryPerPage))
+
+	for k := range c.topLevelChannels {
+		ids = append(ids, k)
+	}
+	sort.Sort(int64Slice(ids))
+	count := 0
+	var ok bool
+	for i, v := range ids {
+		if count == EntryPerPage {
+			break
+		}
+		if v < id {
+			continue
+		}
+		if cn, ok := c.channels[v]; ok {
+			cns = append(cns, cn)
+			count++
+		}
+		if i == len(ids)-1 {
+			ok = true
+			break
+		}
+	}
+	c.mu.RUnlock()
+	if count == 0 {
+		ok = true
+	}
+
+	var t []*ChannelMetric
+	for _, cn := range cns {
+		cim := cn.c.ChannelzMetric()
+		cm := &ChannelMetric{}
+		cm.ChannelData = cim
+		c.mu.RLock()
+		cm.NestedChans = copyMap(cn.nestedChans)
+		cm.SubChans = copyMap(cn.subChans)
+		c.mu.RUnlock()
+		t = append(t, cm)
+	}
+	return t, ok
+}
+
+func (c *channelMap) GetServers(id int64) ([]*ServerMetric, bool) {
+	c.mu.RLock()
+	l := len(c.servers)
+	ids := make([]int64, 0, l)
+	ss := make([]*server, 0, min(l, EntryPerPage))
+	for k := range c.servers {
+		ids = append(ids, k)
+	}
+	sort.Sort(int64Slice(ids))
+	count := 0
+	var ok bool
+	for i, v := range ids {
+		if count == EntryPerPage {
+			break
+		}
+		if v < id {
+			continue
+		}
+		if cn, ok := c.servers[v]; ok {
+			ss = append(ss, cn)
+			count++
+		}
+		if i == len(ids)-1 {
+			ok = true
+			break
+		}
+	}
+	c.mu.RUnlock()
+	if count == 0 {
+		ok = true
+	}
+
+	var s []*ServerMetric
+	for _, cn := range ss {
+		sm := &ServerMetric{}
+		sm.ServerData = cn.s.ChannelzMetric()
+		sm.ID = cn.id
+		sm.RefName = cn.refName
+		c.mu.RLock()
+		sm.ListenSockets = copyMap(cn.listenSockets)
+		c.mu.RUnlock()
+		s = append(s, sm)
+	}
+	return s, ok
+}
+
+func (c *channelMap) GetServerSockets(id int64, startID int64) ([]*SocketMetric, bool) {
+	var cn *server
+	var ok bool
+	c.mu.RLock()
+	if cn, ok = c.servers[id]; !ok {
+		// server with id doesn't exist.
+		c.mu.RUnlock()
+		return nil, true
+	}
+	sm := cn.sockets
+	l := len(sm)
+	ids := make([]int64, 0, l)
+	sks := make([]*normalSocket, 0, min(l, EntryPerPage))
+	for k := range sm {
+		ids = append(ids, k)
+	}
+	sort.Sort((int64Slice(ids)))
+	count := 0
+	ok = false
+	for i, v := range ids {
+		if count == EntryPerPage {
+			break
+		}
+		if v < startID {
+			continue
+		}
+		if cn, ok := c.normalSockets[v]; ok {
+			sks = append(sks, cn)
+			count++
+		}
+		if i == len(ids)-1 {
+			ok = true
+			break
+		}
+	}
+	c.mu.RUnlock()
+	if count == 0 {
+		ok = true
+	}
+	var s []*SocketMetric
+	for _, cn := range sks {
+		sm := &SocketMetric{}
+		sm.SocketData = cn.s.ChannelzMetric()
```

Comment:
delete this line, as SPEC has changed, and don't need the socket's metric. Only id and ref pair is need.